File: w2d5/w2d5/lib/array.rb
require 'byebug'

class Array

  def uniq
    arr = []
    self.each do |val|
      arr << val unless arr.include?(val)
    end

    arr
  end

  def two_sum
    arr = []

    self.each_with_index do |val, idx1|
      ((idx1 + 1)...self.length).each do |idx2|
        if val == -self[idx2]
          arr << [idx1, idx2]
        end
      end
    end

    arr
  end

  def median
    return nil if empty?
    length.even? ? (self[length/2] + self[length/2 - 1])/2 : self[length/2]

  end

  def my_transpose
    transpose_matrix = []
	  self.count.times { transpose_matrix << [] }

    self.each_with_index do |row, row_index|
		# then iterate over each element of the row in the existing matrix
		  row.each_with_index do |row_element, column_index|
			# add the element of each row, to the new column arrays in the transpose matrix
			   transpose_matrix[column_index] << row_element
		  end
	  end


    transpose_matrix
  end

  def stock_picker
  	highest_profit = 0
  	buying_day = nil
  	selling_day = nil

  	self.each_with_index do |buying_pr, buying_d|
  		self.each_with_index do |selling_pr, selling_d|
  			if buying_d < selling_d
  				profit = selling_pr - buying_pr
  				if profit > highest_profit
  					highest_profit = profit
  					buying_day = buying_d
  					selling_day = selling_d
  				end
  			end
  		end
  	end
  	[buying_day, selling_day]
  end


end

#p [1,2,3].my_transpose
#p [2,3,3,2].two_sum

File: w5/w5d4/array.js
Array.prototype.uniq = function () {
  var new_arr = [];

  for (var i = 0; i < this.length; i++) {
    var uniq = true;
    for(var k = 0; k < new_arr.length; k++) {
      if (new_arr[k] === this[i]) {
        uniq = false;
      }
    }
    if (uniq) {
      new_arr.push(this[i])
    }
  }
  return new_arr;
}

var array = [1,2,3,2];
console.log(array.uniq());

Array.prototype.two_sum = function () {
  var new_arr = [];
  for (var i = 0; i < this.length; i++) {
    for (var j = i + 1; j < this.length; j++) {
      if ((this[i] + this[j]) == 0) {
        new_arr.push([i, j]);
      }
    }
  }
  return new_arr;

}

console.log([1, 2, -1, -2].two_sum());

Array.prototype.myTranspose = function () {
  var new_arr = [];
  for (var i = 0; i < this.length; i++) {
    var trans_arr = [];
    for (var j = 0; j < this.length; j++) {
        trans_arr.push(this[j][i]);

    }
    new_arr.push(trans_arr);
  }
  return new_arr;
}

var rows = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8]
  ];

  console.log(rows.myTranspose());

  Array.prototype.stockPicker = function () {
    var new_arr = [];
    var profit = 0;
    for (var i = 0; i < this.length; i++) {
      for (var j = i + 1; j < this.length; j++) {
        if (this[j] - this[i] > profit) {
          new_arr = [i, j];
          profit = this[j] - this[i];
        }
      }
    }
    return new_arr;
  }

var stocks = [ 1, 2, 3, 4 ];

console.log(stocks.stockPicker());
